src/test/java/com/uber/cadence/worker/WorkerStressTests.java

```diff
@@ -0,0 +1,201 @@
+            .getWorkflowClient()
+            .newUntypedWorkflowStub("ActivitiesWorkflow::execute", workflowOptions);
+
+    // Act
+    // This will yeild around 10000 events which is above the page limit returned by the server.
+    WorkflowParams w = new WorkflowParams();
+    w.CadenceSleep = Duration.ofSeconds(0);
+    w.ChainSequence = 50;
+    w.ConcurrentCount = 50;
+    w.PayloadSizeBytes = 10000;
+    w.TaskListName = taskListName;
+
+    workflow.start(w);
+    assertNotNull("I'm done.", workflow.getResult(String.class));
+    wrapper.close();
+  }
+
+  // Todo: refactor TestEnvironment to toggle between real and test service.
+  private class TestEnvironmentWrapper {
+
+    private TestWorkflowEnvironment testEnv;
+    private Worker.Factory factory;
+
+    public TestEnvironmentWrapper(Worker.FactoryOptions options) {
+      if (options == null) {
+        options = new Worker.FactoryOptions.Builder().Build();
+      }
+      factory = new Worker.Factory(DOMAIN, options);
+      TestEnvironmentOptions testOptions =
+          new TestEnvironmentOptions.Builder().setDomain(DOMAIN).setFactoryOptions(options).build();
+      testEnv = TestWorkflowEnvironment.newInstance(testOptions);
+    }
+
+    private Worker.Factory getWorkerFactory() {
+      return useExternalService ? factory : testEnv.getWorkerFactory();
+    }
+
+    private WorkflowClient getWorkflowClient() {
+      return useExternalService ? WorkflowClient.newInstance(DOMAIN) : testEnv.newWorkflowClient();
+    }
+
+    private void close() {
+      factory.shutdown(Duration.ofSeconds(1));
+      testEnv.close();
+    }
+  }
+
+  public static class WorkflowParams {
+
+    public int ChainSequence;
+    public int ConcurrentCount;
+    public String TaskListName;
+    public int PayloadSizeBytes;
+    public Duration CadenceSleep;
+  }
+
+  public interface ActivitiesWorkflow {
+
+    @WorkflowMethod()
+    String execute(WorkflowParams params);
+  }
+
+  public static class ActivitiesWorkflowImpl implements ActivitiesWorkflow {
+    private static final Logger log = LoggerFactory.getLogger(ActivitiesWorkflowImpl.class);
+
+    @Override
+    public String execute(WorkflowParams params) {
+      SleepActivity activity =
+          Workflow.newActivityStub(
+              SleepActivity.class,
+              new ActivityOptions.Builder()
+                  .setTaskList(params.TaskListName)
+                  .setScheduleToStartTimeout(Duration.ofMinutes(1))
+                  .setStartToCloseTimeout(Duration.ofMinutes(1))
+                  .setHeartbeatTimeout(Duration.ofSeconds(20))
+                  .build());
+
+      for (int i = 0; i < params.ChainSequence; i++) {
+        List<Promise<Void>> promises = new ArrayList<>();
+        for (int j = 0; j < params.ConcurrentCount; j++) {
+          byte[] bytes = new byte[params.PayloadSizeBytes];
+          new Random().nextBytes(bytes);
+          Promise<Void> promise = Async.procedure(activity::sleep, i, j, bytes);
```

Comment:
This call doesn't create multiple threads as activity implementation is asynchronous in this case.

Reply:
It does on workflow execution in syncDecisionContext. executeActivity will do it in a new workflowThreadImpl so this will cause the threadpool to be consumed

Reply:
I'm confused. Are you saying that Async.procedure for activity does consume a thread after all your changes? If it is the case it should be fixed asap.

Reply:
Discussed offline